### PR TITLE
[perf] Inventory rework (part 1)

### PR DESCRIPTION
**Pull request by RDIL**
Changes the inventory functions to use a cached map.

### CODE REVIEW

**Review by AnthonyFuller — APPROVED** [review]
Fucking finally.

**Comment by RDIL** [comment]
Some benchmarking we need to do (as performance issues are a problem in this case):

- Currently, createInventory stores the created inventory as a list of inventory objects. This will use a lot of memory. What is the overhead of storing that as a string[] instead, and then translating to actual objects when needed? Which is faster?

File: components/utils.ts [hunk 1]
```diff
@@ -520,50 +520,7 @@ export function addDashesToPublicId(publicId: string): string {
...
-                    result[i] = fastClone(item[i])
-                }
-            }
-        } else {
-            result = item
-        }
-    }
-
-    return result
+    return structuredClone(item)
```
Comment (RDIL) [inline components/utils.ts:523]: New built-in added in Node 17

**Comment by AnthonyFuller** [comment]
> Some benchmarking we need to do (as performance issues are a problem in this case):
> 
>     * Currently, createInventory stores the created inventory as a list of inventory objects. This will use a lot of memory. What is the overhead of storing that as a string[] instead, and then translating to actual objects when needed? Which is faster?

Well isn't it being stored _twice_ due to the config `allunlockables`? So if we can just "unload" that when it's been placed into a map, then we can just eat up the memory usage of the map. We knew the map would be an issue anyway.

**Comment by RDIL** [comment]
@AnthonyFuller there isn't a really good way of unloading it. No matter what we really do, it will still be in memory because of sourcemaps and such

**Comment by AnthonyFuller** [comment]
Then we could just not load them in the first place, and load unlockables separately (in some resource package like challenges and mastery).

**Comment by RDIL** [comment]
We'd also have to rework how plugins add items

**Comment by AnthonyFuller** [comment]
That's fine, plugins are getting a re-work in v7 anyway.

**Pull request merged**